src/otx/algo/visual_prompting/segment_anything.py

```diff
@@ -187,117 +266,36 @@ def forward_inference(
-        for idx, embedding in enumerate(image_embeddings):
-            low_res_masks, iou_predictions = [], []
-            for prompt in [bboxes[idx], points[idx]]:
-                if prompt is None:
-                    continue
-
-                sparse_embeddings, dense_embeddings = self.prompt_encoder(
-                    points=prompt if isinstance(prompt[0], Points) else None,
-                    boxes=prompt if isinstance(prompt, tv_tensors.BoundingBoxes) else None,
-                    masks=None,
-                )
-                _low_res_masks, _iou_predictions = self.mask_decoder(
-                    image_embeddings=embedding.unsqueeze(0),
-                    image_pe=self.prompt_encoder.get_dense_pe(),
-                    sparse_prompt_embeddings=sparse_embeddings,
-                    dense_prompt_embeddings=dense_embeddings,
-                    multimask_output=False,  # when given multiple prompts. if there is single prompt True would be better. # noqa: E501
-                )
-                low_res_masks.append(_low_res_masks)
-                iou_predictions.append(_iou_predictions)
-
-            pred_masks.append(torch.cat(low_res_masks, dim=0))
-            ious.append(torch.cat(iou_predictions, dim=0))
-
-        if self.training:
-            loss_dice = 0.0
-            loss_focal = 0.0
-            loss_iou = 0.0
-
-            num_masks = sum(len(pred_mask) for pred_mask in pred_masks)
-            for pred_mask, gt_mask, iou, ori_shape in zip(pred_masks, gt_masks, ious, ori_shapes):  # type: ignore[arg-type]
-                post_processed_pred_mask = self.postprocess_masks(pred_mask, self.image_size, ori_shape)
-                post_processed_pred_mask = post_processed_pred_mask.sigmoid()
-                post_processed_pred_mask = post_processed_pred_mask.flatten(1)
-                flatten_gt_mask = gt_mask.flatten(1).float()
-
-                # calculate losses
-                loss_dice += self.calculate_dice_loss(post_processed_pred_mask, flatten_gt_mask, num_masks)
-                loss_focal += self.calculate_sigmoid_ce_focal_loss(post_processed_pred_mask, flatten_gt_mask, num_masks)
-                batch_iou = self.calculate_iou(post_processed_pred_mask, flatten_gt_mask)
-                loss_iou += F.mse_loss(iou, batch_iou.unsqueeze(1), reduction="sum") / num_masks
-
-            loss = 20.0 * loss_focal + loss_dice + loss_iou
-
-            return {"loss": loss, "loss_focal": loss_focal, "loss_dice": loss_dice, "loss_iou": loss_iou}
-
-        post_processed_pred_masks: list[Tensor] = []
-        for pred_mask, ori_shape in zip(pred_masks, ori_shapes):
-            post_processed_pred_mask = self.postprocess_masks(pred_mask, self.image_size, ori_shape)
-            post_processed_pred_masks.append(post_processed_pred_mask.squeeze(1).sigmoid())
-        return post_processed_pred_masks, ious
-
     def _embed_points(self, point_coords: Tensor, point_labels: Tensor) -> Tensor:
```

Comment:
It seems that these code execution uses `SegmentAnything` function in almost case. Then how about moving this function and `_embed_masks` to `SegmentAnything`?

Reply:
But it is not used at all in `SegmentAnything`